petstore/src/main/scala/io/finch/petstore/petstoreDb.scala

```diff
@@ -0,0 +1,322 @@
+      val pending: Int = stock(Pending)
+      val adopted: Int = stock(Adopted)
+      Inventory(available, pending, adopted)
+    }
+  )
+
+  /**
+   * POST: Place an [[Order]] for a [[Pet]].
+   * @param order The order object to be placed with the petstore.
+   * @return The autogenerated ID of the order object.
+   */
+  def addOrder(order: Order): Future[Long] =
+    orders.synchronized {
+      order.id match{
+        case Some(_) => Future.exception(InvalidInput("New order should not contain an id"))
+        case None => orders.synchronized{
+          val genId = if (orders.isEmpty) 0 else orders.keys.max + 1
+          orders(genId) = order.copy(id = Some(genId))
+          Future(genId)
+        }
+      }
+    }
+
+  /**
+   * DELETE: Delete purchase [[Order]] by ID
+   * @param id The ID of the order to delete.
+   * @return true if deletion was successful. false otherwise.
+   */
+  def deleteOrder(id: Long): Future[Boolean] = Future.value(
+    orders.synchronized{
+      if (orders.contains(id)) {
+        orders.remove(id)
+        true
+      } else false
+    }
+  )
+
+  /**
+   * GET: Find purchase [[Order]] by ID
+   * @param id The ID of the order to find.
+   * @return The Order object in question.
+   */
+  def findOrder(id: Long): Future[Order] = Future.value(
+    orders.synchronized{
+      orders.getOrElse(id, throw OrderNotFound("Your order doesn't exist! :("))
+    }
+  )
+
+  /**
+   * POST: Create a [[User]].
+   * @param newGuy The User we want to add to the database.
+   * @return The user name of the added User.
+   */
+  def addUser(newGuy: User): Future[String] =
+    users.synchronized {
+      val inputName: String = newGuy.username
+      if(users.values.exists(_.username == inputName)) throw RedundantUsername(s"Username $inputName is already taken.")
+      else{
+        newGuy.id match{
+          case Some(_) => Future.exception(InvalidInput("New user should not contain an id"))
+          case None => users.synchronized{
+            val genId = if (users.isEmpty) 0 else users.keys.max + 1
+            users(genId) = newGuy.copy(id = Some(genId))
+            Future(newGuy.username)
+          }
+        }
+      }
+    }
+
+  /**
+   * GET: Get [[User]] by username, assume all usernames are unique.
+   * @param name The username of the User we want to find.
+   * @return The User in question.
+   */
+  def getUser(name: String): Future[User] = Future.value(
+    users.synchronized{
+      val pickMeIter: Iterable[User] = for{
+        u <- users.values
+        if u.username.equals(name)
+      } yield u
+      if (pickMeIter.size == 0) throw MissingUser("This user doesn't exist!")  else pickMeIter.toSeq(0)
+    }
+  )
+
+  /**
+   * DELETE: Delete a [[User]] by their username.
+   * @param name The username of the User to be deleted.
+   */
+  def deleteUser(name: String): Future[Unit] =
+    users.synchronized {
+      val id: Long = Await.result(getUser(name)).id.get
```

Comment:
Let's prohibit `Await` in this file—it's not a good example to block unnecessarily. This can be rewritten as the following:

```scala
getUser(name).flatMap {
  case Some(id) => users.remove(id).unit
  case None => // Do whatever's appropriate for non-existent users.
}
```